### PR TITLE
BaseCacheManager: Close stream of `getFile` on listener unsubscribed.

### PR DESCRIPTION
**Pull request by yrom**
Use `StreamController` instead of _async*-yield_.
Fixed renefloor/flutter_cached_network_image#216

### CODE REVIEW

File: lib/src/cache_manager.dart [hunk 1]
```diff
@@ -104,35 +104,47 @@ abstract class BaseCacheManager {
...
-        if (webFile != null) {
-          yield webFile;
-        }
-      } catch (e) {
-        if (cacheFile == null) {
-          throw e;
+        cacheFile = await getFileFromCache(url);
+      } catch (e, stack) {
+        // XXX: should we suppress this error?
+        if (!holder.isCanceled) holder.addError(e, stack);
```
Comment (Kavantix) [inline lib/src/cache_manager.dart]: The previous version did not give the error to the stream if the cache could not be retrieved.
Perhaps we could think about a way to communicate this to the developer without returning the error

File: lib/src/cache_manager.dart [hunk 2]
```diff
@@ -104,35 +104,47 @@ abstract class BaseCacheManager {
...
+    return webHelper.downloadFile(url,
         authHeaders: authHeaders, ignoreMemCache: force);
   }
 
   ///Get the file from the cache
-  Future<FileInfo> getFileFromCache(String url) async {
-    return await store.getFile(url);
+  Future<FileInfo> getFileFromCache(String url) {
+    return store.getFile(url);
   }
```
Comment (Kavantix) [inline lib/src/cache_manager.dart]: These changes seem unrelated to the PR

File: lib/src/cache_manager.dart [hunk 3]
```diff
@@ -104,35 +104,47 @@ abstract class BaseCacheManager {
...
   /// The files are returned as stream. First the cached file if available, when the
   /// cached file is too old the newly downloaded file is returned afterwards.
-  Stream<FileInfo> getFile(String url, {Map<String, String> headers}) async* {
-    var cacheFile = await getFileFromCache(url);
-    if (cacheFile != null) {
-      yield cacheFile;
-    }
-    if (cacheFile == null || cacheFile.validTill.isBefore(DateTime.now())) {
+  Stream<FileInfo> getFile(String url, {Map<String, String> headers}) {
+    return _StreamHolder<FileInfo>((holder) async {
```
Comment (Kavantix) [inline lib/src/cache_manager.dart]: Adding the `StreamHolder` seems unnecessary to me.
Instead we could just use a `StreamController` directly.

Comment (Kavantix) [inline lib/src/cache_manager.dart]: Giving the holder a callback also means that `getFileFromCache` will not be called immediately.
Looking at the holder class it seems that this callback is only called when someone starts listening to the stream which is not what the old version did and also not intended i think.

File: lib/src/cache_manager.dart [hunk 4]
```diff
@@ -183,3 +195,36 @@ abstract class BaseCacheManager {
...
+  void add(T event) {
+    _controller.add(event);
+  }
+
+  void addError(dynamic error, [StackTrace trace]) {
+    _controller.addError(error, trace);
+  }
+
+  // for single consumer
+  final StreamController<T> _controller = StreamController<T>();
```
Comment (Kavantix) [inline lib/src/cache_manager.dart]: Try to put all field together at the top of a class

File: lib/src/cache_manager.dart (hunk 3 above)
Comment (Kavantix) [reply lib/src/cache_manager.dart]: @renefloor which behaviour would you prefer?

Comment (yrom) [reply lib/src/cache_manager.dart]: I think that the stream should keep silent except someone is interested.

File: lib/src/cache_manager.dart (hunk 1 above)
Comment (yrom) [reply lib/src/cache_manager.dart]: The previous version was `var cacheFile = await getFileFromCache(url);`. And it would raise an unexpeted-error when `await getFileFromCache(url)` run into exception

File: lib/src/cache_manager.dart (hunk 3 above)
Comment (renefloor) [reply lib/src/cache_manager.dart]: Both opinions make sense. Let me think about it for a moment what the behaviour is that I want. The changes you propose are breaking in some way, and I prefer to prevent breaking changes, but I really like the idea of cancelling the stream when nobody is listening anymore. (I guess in that way it makes sense to only start when somebody is listening.)

**Comment by renefloor** [comment]
@yrom I used your idea of a streamcontroller to better handle errors, but with the original behaviour.
See this pull request: https://github.com/renefloor/flutter_cache_manager/pull/117

I decided to continue download the file, although nobody is listening, because perhaps you want to pre-load the image to use it later. (It is for caching anyhow.) And I made the code a bit cleaner and easier to understand.
